coil-core/src/androidMain/kotlin/coil3/decode/StaticImageDecoder.kt

```diff
@@ -106,40 +108,45 @@ class StaticImageDecoder(
             options: Options,
             imageLoader: ImageLoader,
         ): Decoder? {
-            val source = result.source.imageDecoderSourceOrNull(options) ?: return null
+            val source = result.source.toImageDecoderSource(options) ?: return null
             return StaticImageDecoder(source, result.source, options, parallelismLock)
         }
+    }
+}
 
-        private fun ImageSource.imageDecoderSourceOrNull(options: Options): ImageDecoder.Source? {
-            if (fileSystem === FileSystem.SYSTEM) {
-                val file = fileOrNull()
-                if (file != null) {
-                    return ImageDecoder.createSource(file.toFile())
-                }
-            }
+@InternalCoilApi
+@RequiresApi(28)
+fun ImageSource.toImageDecoderSource(options: Options): ImageDecoder.Source? {
+    if (fileSystem === FileSystem.SYSTEM) {
+        val file = fileOrNull()
+        if (file != null) {
+            return ImageDecoder.createSource(file.toFile())
+        }
+    }
 
-            val metadata = metadata
-            return when {
-                metadata is AssetMetadata -> {
-                    ImageDecoder.createSource(options.context.assets, metadata.filePath)
-                }
-                metadata is ContentMetadata -> try {
+    val metadata = metadata
+    return when {
```

Comment:
`when (val metadata = metadata)`
See https://github.com/Kotlin/KEEP/blob/master/proposals/val-in-when-subject.md

Reply:
Unfortunately we can't use the when subject since we have extra `&&` statements (e.g. `metadata is ResourceMetadata && metadata.packageName == options.context.packageName`).

Reply:
Friendly remind: maybe we can use guards later.
https://github.com/Kotlin/KEEP/blob/guards/proposals/guards.md